### PR TITLE
fixes for MacOS and new CI to prevent future breaks

### PR DESCRIPTION
**Pull request by jeanbez**
Fixes for MacOS by correctly handle finding `UUID_LIBRARY` (the current code was mixing the library provided by the system, but no the headers, and getting partially overwritten by a local user installation). Some tests were also breaking due to wrong memory allocation. For MacOS we also need to run mercury with `sockets` instead of `tcp`.

We also include a new CI check where we compile PDC with MacOS, but we do not run tests due to some issues in Github provided runners (https://github.com/actions/runner-images/issues/3885)

### CODE REVIEW

File: .github/workflows/dependencies-macos.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+
+set -eu -o pipefail
+
+brew install open-mpi autoconf automake libtool ossp-uuid
+
+# libfabric
+wget https://github.com/ofiwg/libfabric/archive/refs/tags/v1.12.1.tar.gz
```
Comment (houjun) [inline .github/workflows/dependencies-macos.sh]: We can upgrade libfabric to at least v1.15.2

File: .github/workflows/dependencies-macos.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+./autogen.sh
+./configure --disable-usnic --disable-mrail --disable-rstream --disable-perf --disable-efa --disable-psm2 --disable-psm --disable-verbs --disable-shm --disable-static --disable-silent-rules
+make -j2 && sudo make install
+make check
+cd ..
+
+# Mercury
+git clone --recursive https://github.com/mercury-hpc/mercury.git
+cd mercury
+git checkout v2.1.0
```
Comment (houjun) [inline .github/workflows/dependencies-macos.sh]: We can upgrade mercury to v2.2.0

File: src/commons/utils/include/pdc_malloc.h [hunk 3]
```diff
@@ -76,8 +76,8 @@ void *PDC_free(void *mem);
  */
 size_t PDC_get_global_mem_usage();
 
-#define PDC_MALLOC(t)    (t *)PDC_malloc(sizeof(t))
-#define PDC_CALLOC(c, t) (t *)PDC_calloc(c, sizeof(t))
+#define PDC_MALLOC(t)    (t *)PDC_malloc(sizeof(t) + 1)
+#define PDC_CALLOC(c, t) (t *)PDC_calloc(c, sizeof(t) + 1)
```
Comment (houjun) [inline src/commons/utils/include/pdc_malloc.h]: Why +1 here?

Comment (jeanbez) [reply src/commons/utils/include/pdc_malloc.h]: This was causing the memory segfault in strdup for the tags tests in MacOS. Apparently, there was not enough memory allocated for those which caused the error.

File: .github/workflows/dependencies-macos.sh (hunk 1 above)
Comment (jeanbez) [reply .github/workflows/dependencies-macos.sh]: Done!

File: src/commons/utils/include/pdc_malloc.h (hunk 3 above)
Comment (zhangwei217245) [reply src/commons/utils/include/pdc_malloc.h]: - Any clear explanation on why +1 is the fix? 
- Is this a fix only for MacOS? 
- If MacOS needs this +1 fix, what about Linux? 
- Plus, the '+1' fix on calloc call will actually result in 'c' more bytes allocated each time. Any negative impact?

Comment (jeanbez) [reply src/commons/utils/include/pdc_malloc.h]: We found the origin of the issue. This change has been rolledback.

**Review by houjun — APPROVED** [review]

File: CMake/FindUUID.cmake [hunk 4]
```diff
@@ -1,36 +1,41 @@
...
-mark_as_advanced(UUID_INCLUDE_DIR UUID_LIBRARY)
\ No newline at end of file
+  PATHS /usr/local/include /usr/include
+)
+
+if (UUID_LIBRARY AND UUID_INCLUDE_DIR)
+  set(UUID_LIBRARIES ${UUID_LIBRARY})
+  set(UUID_FOUND "YES")
+else ()
+  set(UUID_FOUND "NO")
```
Comment (zhangwei217245) [inline CMake/FindUUID.cmake]: Similarly, FALSE instead of "NO“？

File: CMake/FindUUID.cmake [hunk 5]
```diff
@@ -1,36 +1,41 @@
...
-                                  UUID_INCLUDE_DIR UUID_LIBRARY)
-
-mark_as_advanced(UUID_INCLUDE_DIR UUID_LIBRARY)
\ No newline at end of file
+  PATHS /usr/local/include /usr/include
+)
+
+if (UUID_LIBRARY AND UUID_INCLUDE_DIR)
+  set(UUID_LIBRARIES ${UUID_LIBRARY})
+  set(UUID_FOUND "YES")
```
Comment (zhangwei217245) [inline CMake/FindUUID.cmake]: Shall we set it to TRUE instead of "YES"?

**Review by zhangwei217245 — CHANGES_REQUESTED** [review]

**Comment by zhangwei217245** [comment]
Tested on my Mac with the guide provided by Jean Luca, no issue with compilation. the first 66 ctest ran smoothly, and like the situation in a linux PC, the 67th test takes a long time. 

some flags used for installing libfabric and mercury on MacOS are different, I suggest we should open **a dedicated subsection in the installation guide document to provide clear commands for Mac users.** 

Below are the flags provided by Jean Luca and they are working  on my Mac:

libfabric: ./configure --prefix=<install_path> CFLAG=-O2 --enable-sockets=yes --enable-tcp=yes --enable-udp=yes --enable-rxm=yes 
Mercury: -DBUILD_SHARED_LIBS=ON -DBUILD_TESTING=ON -DNA_USE_OFI=ON -DNA_USE_SM=OFF -DNA_OFI_TESTING_PROTOCOL=sockets
PDC: -DBUILD_MPI_TESTING=ON -DBUILD_SHARED_LIBS=ON -DBUILD_TESTING=ON -DCMAKE_INSTALL_PREFIX=$PDC_DIR -DPDC_ENABLE_MPI=ON -DMERCURY_DIR=$MERCURY_DIR -DCMAKE_C_COMPILER=mpicc -DMPI_RUN_CMD=mpiexec ../
  and also    export HG_TRANSPORT="sockets" (before ctest)


Also, documentation should be added about authorizing all mercury and  PDC ctests in MacOS to avoid dialog box being popped continuously.

**Comment by zhangwei217245** [comment]
Test #67 and later just timed out. 

image

Would it be possible if we can setup a shorter timeout limit for these ctests?

**Comment by jeanbez** [comment]
> Would it be possible if we can setup a shorter timeout limit for these ctests?

There is a global way to set the timeout directly in CMake, we can either set the timeout for each test or you can change the way you call `ctest` to `ctest --timeout 60` for a 60 seconds timeout. I believe the second approach would be better than manually adding a timeout for each of the >100 tests we have.

**Comment by jeanbez** [comment]
Updated documentation and the CI can now run in macOS runners after GitHub's latest updates on their support.

**Review by houjun — APPROVED** [review]

**Review by zhangwei217245 — APPROVED** [review]

**Pull request merged**